From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Austen Adler <agadler@austenadler.com>
Date: Sun, 4 Feb 2024 00:15:41 -0500
Subject: [PATCH] 6195: Add Copy Special option for byte source offset

Add clipboard option for byte source offset

Fix docs for byte source offset clipboard

Fix variable names for byte source offset clipboard

Copy <NO_OFFSET> if address is not backed by a file

Change final BYTE_SOURCE_OFFSET_TYPE name to be more consistent
---
 .../help/topics/ClipboardPlugin/Clipboard.htm |  8 +++--
 .../CodeBrowserClipboardProvider.java         | 31 +++++++++++++++++++
 2 files changed, 37 insertions(+), 2 deletions(-)

diff --git a/Ghidra/Features/Base/src/main/help/help/topics/ClipboardPlugin/Clipboard.htm b/Ghidra/Features/Base/src/main/help/help/topics/ClipboardPlugin/Clipboard.htm
index 0bc4b2e13..addd6cb72 100644
--- a/Ghidra/Features/Base/src/main/help/help/topics/ClipboardPlugin/Clipboard.htm
+++ b/Ghidra/Features/Base/src/main/help/help/topics/ClipboardPlugin/Clipboard.htm
@@ -181,8 +181,12 @@
         <LI><B>Address w/ Offset</B> - Copies the address at the cursor or each address in the 
         current selection.   The text is formatted to show the offset from the entry point of the
         function, for example: <CODE>main + 0x2</CODE></LI>
-        
-	<LI><B>GhidraURL</B> <A name="GhidraURL"></A> - Creates a GhidraURL for the address under the 
+
+        <LI><B>Byte Source Offset</B> - Copies the byte source offset from the start of the file for
+        each address in the current selection. If the address is not backed by a file,
+        <CODE>%3CNO_OFFSET%3E</CODE> is copied.</LI>
+
+	<LI><B>GhidraURL</B> <A name="GhidraURL"></A> - Creates a GhidraURL for the address under the
 	cursor then copies that URL to the clipboard.</LI>
       </UL>
 
diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/clipboard/CodeBrowserClipboardProvider.java b/Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/clipboard/CodeBrowserClipboardProvider.java
index afe37c17a..425376d38 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/clipboard/CodeBrowserClipboardProvider.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/clipboard/CodeBrowserClipboardProvider.java
@@ -44,10 +44,12 @@ import ghidra.app.util.viewer.listingpanel.ListingModel;
 import ghidra.framework.cmd.Command;
 import ghidra.framework.model.DomainFile;
 import ghidra.framework.plugintool.PluginTool;
+import ghidra.program.database.mem.AddressSourceInfo;
 import ghidra.program.database.symbol.CodeSymbol;
 import ghidra.program.database.symbol.FunctionSymbol;
 import ghidra.program.model.address.*;
 import ghidra.program.model.listing.*;
+import ghidra.program.model.mem.Memory;
 import ghidra.program.model.symbol.*;
 import ghidra.program.util.*;
 import ghidra.util.Msg;
@@ -65,6 +67,8 @@ public class CodeBrowserClipboardProvider extends ByteCopier
 		new ClipboardType(DataFlavor.stringFlavor, "Address");
 	public static final ClipboardType ADDRESS_TEXT_WITH_OFFSET_TYPE =
 		new ClipboardType(DataFlavor.stringFlavor, "Address w/ Offset");
+	public static final ClipboardType BYTE_SOURCE_OFFSET_TYPE =
+		new ClipboardType(DataFlavor.stringFlavor, "Byte Source Offset");
 	public static final ClipboardType CODE_TEXT_TYPE =
 		new ClipboardType(DataFlavor.stringFlavor, "Formatted Code");
 	public static final ClipboardType LABELS_COMMENTS_TYPE =
@@ -94,6 +98,7 @@ public class CodeBrowserClipboardProvider extends ByteCopier
 		list.add(CPP_BYTE_ARRAY_TYPE);
 		list.add(ADDRESS_TEXT_TYPE);
 		list.add(ADDRESS_TEXT_WITH_OFFSET_TYPE);
+		list.add(BYTE_SOURCE_OFFSET_TYPE);
 
 		return list;
 	}
@@ -221,6 +226,9 @@ public class CodeBrowserClipboardProvider extends ByteCopier
 		else if (copyType == ADDRESS_TEXT_WITH_OFFSET_TYPE) {
 			return copySymbolString(monitor);
 		}
+		else if (copyType == BYTE_SOURCE_OFFSET_TYPE) {
+			return copyByteSourceOffset(monitor);
+		}
 		else if (copyType == CODE_TEXT_TYPE) {
 			return copyCode(monitor);
 		}
@@ -382,6 +390,29 @@ public class CodeBrowserClipboardProvider extends ByteCopier
 		return createStringTransferable(StringUtils.join(strings, "\n"));
 	}
 
+	private	Transferable copyByteSourceOffset(TaskMonitor monitor) {
+		AddressSetView addrs = getSelectedAddresses();
+		Memory currentMemory = currentProgram.getMemory();
+		List<String> strings = new ArrayList<>();
+		AddressIterator	addresses = addrs.getAddresses(true);
+		while (addresses.hasNext() && !monitor.isCancelled()) {
+			AddressSourceInfo addressSourceInfo = currentMemory.getAddressSourceInfo(addresses.next());
+			if (addressSourceInfo != null) {
+				long fileOffset	= addressSourceInfo.getFileOffset();
+				String fileOffsetString;
+
+				if (fileOffset == -1) {
+					fileOffsetString = "<NO_OFFSET>";
+				} else {
+					fileOffsetString = String.format("%x", addressSourceInfo.getFileOffset());
+				}
+
+				strings.add(fileOffsetString);
+			}
+		}
+		return createStringTransferable(StringUtils.join(strings, "\n"));
+	}
+
 	protected Transferable copyCode(TaskMonitor monitor) {
 
 		AddressSetView addressSet = getSelectedAddresses();
-- 
2.43.0

